docs/efficient_gas_calculation_algorithm.md

```diff
@@ -0,0 +1,156 @@
+# Efficient gas calculation algorithm for EVM
+
+This article describes how to efficiently calculate gas and check stack requirements
+for Ethereum Virtual Machine (EVM) instructions.
+
+
+## Instructions metadata
+
+Let's start by defining some basic and universal instructions' parameters.
+
+1. Base gas cost.
+   
+   This is the static gas cost of instructions. Some instructions may have 
+   additional cost depending on their operand values - these have to be
+   handled individually during the instruction execution.
+   
+2. Stack height requirement.
+
+   This is the minimum stack height (number of items on the stack) 
+   required for the instruction execution.
+   
+3. Stack height change.
+
+   This is difference of the stack height before and after the instruction 
+   execution. Can be negative if the instruction pops more items than pushes.
+   
+Examples:
+
+| opcode  | base gas cost | stack height requirement | stack height change |
+| ------- | ------------- | ------------------------ | ------------------- |
+| ADD     | 3             | 2                        | -1                  |
+| EXP     | 50            | 2                        | -1                  |
+| DUP4    | 3             | 4                        | 1                   |
+| SWAP1   | 3             | 2                        | 0                   |
+| ADDRESS | 2             | 0                        | 1                   |
+| CALL    | 700           | 7                        | -6                  |
+
+
+## Basic blocks
+
+A _basic block_ is a sequence of instructions that are executed "straight-line"
+without being interrupted by jumps. I.e. they are nodes in the _control flow graph_.
+
+The name "basic block" has been taken from LLVM. Other names are: just "block"
+in wasm or "subroutine" in subroutine-threaded interpreters.
+
+In EVM there are simple rules to identify basic block boundaries:
+
+1. The following instructions _end_ a basic block:
+   - `JUMP`,
+   - `JUMPI`,
+   - `STOP`,
+   - `RETURN`,
+   - `REVERT`,
+   - `SELFDESTRUCT`,
+   - an instruction directly preceding `JUMPDEST`.
+
+2. The following instructions _start_ a basic block:
+   - `JUMPDEST`,
+   - the first instruction in the code,
+   - an instruction directly after an instruction that has ended previous basic block.
+
+
+## Algorithm
+
+The algorithm for calculating gas and checking stack requirements precomputes
+the values for basic blocks and during execution the checks are done once per block.
+
+### Collecting requirements for basic blocks
+
+For a basic block we need to collect following information:
+
+- total base gas required by instructions,
+- the stack height required (the minimum stack height needed to execute 
+  all instructions in the block),
+- ~~the start-to-end stack height change~~,
+- the relative maximum stack height.
```

Comment:
What exactly is this?

Reply:
The maximum stack height increase in the basic block relatively to the stack height on which the block execution starts. If `stack_max` is 5 and the block execution starts at stack height 10 the absolute stack height will reach 15.

Reply:
So maybe: "The maximum stack height relative to the stack height at block start"? Also is that per instruction in the block or is it a single value?

Reply:
Per block.

Reply:
Maybe "the maximum stack height change over the block"?
I would also call it `stack_change_max` in the pseudocode

Reply:
@axic proposed `stack_max_growth`, and I renamed is like that in the code.